### PR TITLE
Use Gemfile groups to manage production vs dev dependencies

### LINKED ISSUE #227: Use Gemfile groups to manage production vs dev dependencies
## Is your feature request related to a problem? Please describe.

Currently, Gem dependencies are managed by using a special "GEMFILE TRIM MARKER" text block in the Gemfile to delineate
production dependencies from development and test dependencies. This trim marker was then used by a `sed` command to create temporary copies of the Gemfile with portions trimmed as required for production images, and then the original Gemfile is restored.

Gem dependency management would be much cleaner using the `development` and `test` Gemfile groups that are already defined, along with the `--with` and `--without` flags for bundle operations.

## Describe the solution you would like

The use of the special trim string is eliminated, and replace with the use of Gemfile dependency groups.

### PR DESCRIPTION
**Pull request by diverdane**
### What does this PR do?

This change uses Gemfile groups to manage dependencies in terms
of which dependencies will be included in:
- Production images
- Development images
- Test images

Before this change, dependencies were managed and separated by including a
special "GEMFILE TRIM MARKER" text block in the Gemfile to delineate
production dependencies from development and test dependencies.
This trim marker was then used by a `sed` command to create temporary
copies of the Gemfile with portions trimmed as required for
production images.

A couple of other changes included:
- The Dockerfile.test is changed so that the build of this image is from scratch
   (from Ruby base image) rather than being based on the production service broker image.
   This is done because the base production service broker image includes a COPY of the
   repository, whereas the Dockerfile.test image will assume that the repository
   is volume mounted (so no need to rebuild Docker layers for the test image
   every time a change is made to source code).
- The Dockerfile.test.dev file is eliminated, since images made from Dockerfile.test
    are equivalent.

### What ticket does this PR close?
Resolves #227

#### Change log
- [ ] The CHANGELOG has been updated, or
- [x] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [x] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [x] This PR does not require updating any documentation

### CODE REVIEW

File: Dockerfile.dev [hunk 1]
```diff
@@ -1,16 +1,25 @@
...
                        libpq-dev \
                        zip
 
-RUN mkdir -p /app
-WORKDIR /app
-
 COPY Gemfile \
-     Gemfile.lock /app/
+     Gemfile.lock /tmp/
+WORKDIR /tmp
```
Comment (BradleyBoutcher) [inline Dockerfile.dev:17]: We should set this at the top of the file

File: Dockerfile.test [hunk 2]
```diff
@@ -11,8 +21,17 @@ RUN wget -q https://packages.cloudfoundry.org/debian/cli.cloudfoundry.org.key &&
     apt-get update && \
     apt-get install -y cf-cli
 
-# Install all the ruby dependencies (including test/dev ones)
+# Install all the ruby dependencies (including test ones)
+COPY Gemfile \
+     Gemfile.lock /tmp/
+WORKDIR /tmp
```
Comment (BradleyBoutcher) [inline Dockerfile.test:27]: Same as above comment

File: dev/test_unit [hunk 3]
```diff
@@ -10,6 +12,7 @@ docker run \
     -e CONJUR_AUTHN_LOGIN=admin \
     -e SECURITY_USER_NAME=TEST_USER_NAME \
     -e SECURITY_USER_PASSWORD=TEST_USER_PASSWORD \
+    -v "$(pwd)"/..:/app \
```
Comment (BradleyBoutcher) [inline dev/test_unit:15]: Out of curiosity, is there any benefit or drawbacks from using this, over having `cd "$(dirname "$0")/.."` at the top of the file, and mounting `$(pwd)` directly here?

Comment (diverdane) [reply dev/test_unit:15]: I'm thinking that tests scripts for a given repo will likely share code/utils, so maybe it's good to default them to running out of the directory that they're in?

Comment (BradleyBoutcher) [reply dev/test_unit:15]: Good point!

**Comment by izgeri** [comment]
@diverdane @BradleyBoutcher can you please manually confirm that the ZIP artifact includes the vendored dependencies and DOES NOT include the dev/test dependencies?

I think we had implemented the workaround because the `--without` flag wasn't functioning as expected - just want to be very careful with this PR and confirm the actual contents of the ZIP artifact before approving.

**Comment by BradleyBoutcher** [comment]
@izgeri That reminds me, should we be saving the artifact for the service broker built by the pipeline?

**Comment by BradleyBoutcher** [comment]
I can confirm that the `vendor/cache` directory bundled inside the ZIP does not contain any of the dependencies in the `development` groups

**Comment by diverdane** [comment]
> I think we had implemented the workaround because the `--without` flag wasn't functioning as expected - just want to be very careful with this PR and confirm the actual contents of the ZIP artifact before approving.

This may have had to do with a HORRIBLE aspect of the `bundle install ...` for bundle version < 2.0. For bundle version < 2.0, instead of things be stateless (and idempotent), bundle "remembers" any previous use of the `--without` flag to exclude a given group. So the next invocation of `bundle`, we have to explicitly use `--with` of to force bundle to forget the previous `--without`.

**Review by BradleyBoutcher — APPROVED** [review]

**Pull request merged**